### PR TITLE
[Example] Enhance stable_fluid.py with BFECC, RK2, RK3 and Taichi's new syntax

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
Btw, who wrote this cool example?

### CODE REVIEW

File: examples/stable_fluid.py [hunk 1]
```diff
@@ -95,83 +153,87 @@ def apply_impulse(vf: ti.template(), dyef: ti.template(),
...
+        vr = sample(vf, i + 1, j).y
+        vb = sample(vf, i, j - 1).x
+        vt = sample(vf, i, j + 1).x
+        vc = sample(vf, i, j)
+        velocity_curls[i, j] = (vr - vl - vt + vb) * 0.5
 
 
 @ti.kernel
 def pressure_jacobi(pf: ti.template(), new_pf: ti.template()):
+    ti.cache_read_only(pf)
```
Comment (archibate) [inline examples/stable_fluid.py:203]: Btw, can we document this feature somewhere? I'm puzzled, despite it brings a bit of FPS improvements, and I guess I'm using it correct...

File: python/taichi/lang/__init__.py [hunk 2]
```diff
@@ -211,22 +211,22 @@ def init(arch=None,
...
 
 
-def cache_read_only(v):
-    taichi_lang_core.cache(1, v.ptr)
+def cache_read_only(*args):
+    for v in args:
+        taichi_lang_core.cache(1, v.loop_range().ptr)
 
 
 def assume_in_range(val, base, low, high):
```
Comment (archibate) [inline python/taichi/lang/__init__.py:227]: What is this function used for? Can we have a section `performance.rst` in `Advanced programming` explaining these cool APIs? Or they're just incomplete or even buggy for now?

File: examples/stable_fluid.py [hunk 3]
```diff
@@ -1,33 +1,33 @@
 # References:
 # http://developer.download.nvidia.com/books/HTML/gpugems/gpugems_ch38.html
 # https://github.com/PavelDoGreat/WebGL-Fluid-Simulation
+# https://www.bilibili.com/video/BV1ZK411H7Hc?p=4
+# https://github.com/ShaneFX/GAMES201/tree/master/HW1
```
Comment (k-ye) [inline examples/stable_fluid.py:5]: Broken?

Comment (archibate) [reply examples/stable_fluid.py:5]:
```suggestion
# https://github.com/ShaneFX/GAMES201/tree/master/HW01
```

**Pull request merged**